runtime/jcl/common/jdk_internal_misc_ScopedMemoryAccess.cpp

```diff
@@ -34,11 +36,76 @@ Java_jdk_internal_misc_ScopedMemoryAccess_registerNatives(JNIEnv *env, jclass cl
 {
 }
 
+static UDATA
+closeScope0FrameWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState)
+{
+	if (JNI_FALSE == *(jboolean *)walkState->userData2) {
+		/* scope has been found */
+		return J9_STACKWALK_STOP_ITERATING;
+	}
+	return J9_STACKWALK_KEEP_ITERATING;
+}
+
+static void
+closeScope0OSlotWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState, j9object_t *slot, const void *stackLocation)
+{
+	J9Method *method = walkState->method;
+	if (NULL != method) {
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		if (NULL != romMethod && J9ROMMETHOD_HAS_EXTENDED_MODIFIERS(romMethod)) {
+			U_32 extraModifiers = getExtendedModifiersDataFromROMMethod(romMethod);
+			if (J9ROMMETHOD_HAS_SCOPED_ANNOTATION(extraModifiers)) {
+				if (*slot == J9_JNI_UNWRAP_REFERENCE(walkState->userData1)) {
+					*(jboolean *)walkState->userData2 = JNI_FALSE;
+				}
+			}
+		}
+	}
+}
+
+/**
+ * For each thread, walk Java stack and look for the scope instance. Methods that can take and access a scope
+ * instance are marked with the "@Scoped" extended modifier. If the scope instance is found in a method, that
+ * method is accessing the memory segment associated with the scope and thus closing the scope will fail.
+ */
 jboolean JNICALL
 Java_jdk_internal_misc_ScopedMemoryAccess_closeScope0(JNIEnv *env, jobject instance, jobject scope, jobject exception)
 {
-	Assert_JCL_unimplemented();
-	return JNI_FALSE;
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	jboolean scopeNotFound = JNI_TRUE;
```

Comment:
I find this to be a little backwards - why not have the variable be `scopeFound`?

Reply:
https://github.com/eclipse-openj9/openj9/pull/12857#discussion_r651261049

Reply:
Fair enough.